core/block_flyout_inflater.ts

```diff
@@ -0,0 +1,200 @@
+import type {IFlyout} from './interfaces/i_flyout.js';
+import type {IFlyoutInflater} from './interfaces/i_flyout_inflater.js';
+import type {IBoundedElement} from './interfaces/i_bounded_element.js';
+import {BlockSvg} from './block_svg.js';
+import type {WorkspaceSvg} from './workspace_svg.js';
+import * as utilsXml from './utils/xml.js';
+import * as eventUtils from './events/utils.js';
+import * as Xml from './xml.js';
+import * as blocks from './serialization/blocks.js';
+import * as common from './common.js';
+import * as registry from './registry.js';
+import {MANUALLY_DISABLED} from './constants.js';
+import type {Abstract as AbstractEvent} from './events/events_abstract.js';
+import type {BlockInfo} from './utils/toolbox.js';
+import * as browserEvents from './browser_events.js';
+
+/**
+ * The language-neutral ID for when the reason why a block is disabled is
+ * because the workspace is at block capacity.
+ */
+const WORKSPACE_AT_BLOCK_CAPACITY_DISABLED_REASON =
+  'WORKSPACE_AT_BLOCK_CAPACITY';
+
+export class BlockFlyoutInflater implements IFlyoutInflater {
+  protected permanentlyDisabledBlocks = new Set<BlockSvg>();
+  protected listeners = new Map<string, browserEvents.Data[]>();
+  protected flyoutWorkspace?: WorkspaceSvg;
+  protected flyout?: IFlyout;
+  private capacityFilter: (e: AbstractEvent) => void;
+
+  constructor() {
+    this.capacityFilter = (event: AbstractEvent) => {
+      if (
+        !this.flyoutWorkspace ||
+        (event &&
+          !(
+            event.type === eventUtils.BLOCK_CREATE ||
+            event.type === eventUtils.BLOCK_DELETE
+          ))
+      )
+        return;
+
+      this.flyoutWorkspace.getTopBlocks(false).forEach((block) => {
+        if (!this.permanentlyDisabledBlocks.has(block)) {
+          this.updateStateBasedOnCapacity(block);
+        }
+      });
+    };
+  }
+
+  load(state: Object, flyoutWorkspace: WorkspaceSvg): IBoundedElement {
+    this.setFlyoutWorkspace(flyoutWorkspace);
+    this.flyout = flyoutWorkspace.targetWorkspace?.getFlyout() ?? undefined;
+    const block = this.createBlock(state as BlockInfo, flyoutWorkspace);
+
+    if (!block.isEnabled()) {
+      // Record blocks that were initially disabled.
+      // Do not enable these blocks as a result of capacity filtering.
+      this.permanentlyDisabledBlocks.add(block);
+    } else {
+      this.updateStateBasedOnCapacity(block);
+    }
+
+    // Mark blocks as being inside a flyout.  This is used to detect and
+    // prevent the closure of the flyout if the user right-clicks on such
+    // a block.
+    block.getDescendants(false).forEach((b) => (b.isInFlyout = true));
+    this.addBlockListeners(block);
+
+    return block;
+  }
+
+  createBlock(blockDefinition: BlockInfo, workspace: WorkspaceSvg): BlockSvg {
+    let block;
+    if (blockDefinition['blockxml']) {
```

Comment:
Prefer to access properties on `blockDefinition` with `blockDefinition.propertyName` instead of `blockDefinition['propertyname']`

Reply:
Here and throughout.

Reply:
As discussed offline dot access gets minified/obfuscated so this would be unsafe when handling JSON.